ce/amf/metadata-file.ts

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+
+import { extname } from 'path';
+import { Document, LineCounter, parseDocument, YAMLMap } from 'yaml';
+import { i } from '../i18n';
+import { ErrorKind } from '../interfaces/error-kind';
+import { Profile } from '../interfaces/metadata/metadata-format';
+import { ValidationError } from '../interfaces/validation-error';
+import { Session } from '../session';
+import { Uri } from '../util/uri';
+import { BaseMap } from '../yaml/BaseMap';
+import { Coerce } from '../yaml/Coerce';
+import { toYAML } from '../yaml/yaml';
+import { Contacts } from './contact';
+import { Demands } from './demands';
+import { DocumentContext } from './document-context';
+import { GlobalSettings } from './global-settings';
+import { Info } from './info';
+import { Installs } from './installer';
+import { Registries } from './registries';
+import { Requires } from './Requires';
+import { Settings } from './settings';
+
+
+export class MetadataFile extends BaseMap implements Profile {
+  readonly context: DocumentContext;
+  session!: Session;
+  /** @internal */
+  constructor(protected readonly document: Document.Parsed, public readonly filename: string, public lineCounter: LineCounter) {
+    super(<YAMLMap<string, any>><any>document.contents);
+    this.context = <DocumentContext>{
+      filename,
+      lineCounter,
+    };
+  }
+
+  async init(session: Session): Promise<MetadataFile> {
```

Comment:
the only calls to the constructor are thru the factory functions. 

Buuut, I can make it so that the factories are static methods in the class instead.